Add CrUX History API documentation (#5087)
GoogleChrome/developer.chrome.com: site/en/blog/chrome-ux-report-history-api/index.md

---
layout: layouts/blog-post.njk

> title: History web performance data via the CrUX API

Reviewer: ```suggestion
title: Introducing historical web performance data via the CrUX History API
```

description: >

>   History web performance data via the CrUX API.

Reviewer: ```suggestion
  Learn about the new CrUX History API and how to use it to track user experience trends.
```

## Querying the Daily CrUX API

To recap from [our previous article](/blog/chrome-ux-report-api/), you can get a snapshot of the field data for a particular origin from our API in this way:

```shell
API_KEY="[YOUR_API_KEY]"
curl "https://chromeuxreport.googleapis.com/v1/records:queryRecord?key=$API_KEY" --header 'Content-Type: application/json' --data '{"origin": "https://web.dev"}'
```

```json
{
  "record": {
    "key": {
      "origin": "https://web.dev"
    },
    "metrics": {
      "largest_contentful_paint": {
        "histogram": [{
          "start": 0, "end": 2500, "density": 0.9192
        }, {
          "start": 2500, "end": 4000, "density": 0.0513
        }, {
          "start": 4000, "density": 0.0294
        }],
        "percentiles": {
          "p75": 1303
        }
      }
      // ...
    },
    "collectionPeriod": {
      "firstDate": { "year": 2022, "month": 12, "day": 27 },
      "lastDate": { "year": 2023, "month": 1, "day": 23 }
    }
  }
}
```

This snapshot includes histogram density values and percentile values for a particular 28-day collection period, in this case, from December 27, 2002, to January 23, 2023.

> ## Querying the History CrUX API

Reviewer: ```suggestion
## Querying the CrUX History API
```

> This snapshot includes histogram density values and percentile values for a particular 28-day collection period, in this case, from December 27, 2002, to January 23, 2023.

Reviewer: ```suggestion
This snapshot includes histogram density values and percentile values for a particular 28-day collection period, in this case, from December 27, 2022, to January 23, 2023.
```

{% Img src="image/kd0zRXNeKScEzrwMmhY1NNvIzeM2/3izH0o8vNvZqXGMWYCFJ.png", alt="Stacked bar chart showing how the relative proportions of good, needs improvement and poor changes over time.", width="678", height="353" %}

The x-axis is set up as the `lastDate` for each collection period. This time, however, the y-axis is the percentage of page loads falling into a specific range for the INP metric, shown through a stacked bar chart. The p75 chart provides a quick overview, and within a single chart, it’s relatively easy to add multiple metrics, or show lines for both `PHONE` and `DESKTOP`. The tri-bin chart gives a sense of the distribution of metric values measured during each collection period.

For example, even though the p75 chart suggests that [https://web.dev](https://web.dev) had almost acceptable INP values during the observation period, the tri-bin chart shows that for a small fraction of page loads, the INP was actually poor. In both charts, it’s relatively easy to infer that there was a performance regression which started toward the end of October, and was fixed by mid-November.

> To generate such charts yourself, we have created an example [Colab](https://colab.research.google.com/). A Colab, or 'Colaboratory', allows you to write and execute Python in your browser. The [CrUX History API Colab](https://colab.research.google.com/github/GoogleChrome/CrUX/blob/main/colab/crux-history-api.ipynb) [[source]](https://github.com/GoogleChrome/CrUX/blob/main/colab/crux-history-api.ipynb) uses Python to make calls to the API and chart the data.

Reviewer 2: in your browser -> from within your browser (the python runs in some cloud process)

> {% Aside %}

Reviewer 2: Looks like the warning is gone so we can remove this aside.

In other words, interpreting the last time series entries in the example for [https://web.dev](https://web.dev), it was found that from August 14, 2022 until September 10, 2022, 91.87% of page loads had LCP values smaller than 2500ms, 5.27% had values between 2500ms and 4000ms, and 2.85% had values greater than 4000ms.

Similarly, there’s a time series for the p75 values: the LCP p75 for August 14, 2022 until September 10, 2022 was 1377. This means that, for this collection period, 75% of user experiences had an LCP of less than 1377 ms, and 25% of user experiences had an LCP greater than 1377 ms.

While the example only lists 6 time series entries and collection periods, responses from the API currently provide 25 time series entries; since the end dates for each of these collection periods are Saturdays that are 7 days apart, this covers 6 months.

{% Aside %}

> Unlike the daily API, the historical API is only available once a week. However, like the daily API, each value in the historical API is measured from the previous 28 days. This means the historical API should represent the same data that was available via the daily API for the given date—it is just now available to query retrospectively. However, it does mean the historical API contains overlapping periods of data but—as a trend analysis—it is useful to track the change in metrics over time.

Reviewer 2: maybe: is only available once a week -> updates only once a week

> Unlike the daily API, the historical API is only available once a week. However, like the daily API, each value in the historical API is measured from the previous 28 days. This means the historical API should represent the same data that was available via the daily API for the given date—it is just now available to query retrospectively. However, it does mean the historical API contains overlapping periods of data but—as a trend analysis—it is useful to track the change in metrics over time.

Reviewer 2: It looks like all the "CrUX History API" mentions reverted back to "historical API". I think we had this standardized to "CrUX History API" earlier and maybe that's nicer for consistency? Maybe CTRL+F across the file to fix these up?

Author: Fixed.

 --data '{"url": "https://web.dev/blog/"}'
```

Page-level (and origin-level) historical data is subject to [the same eligibility requirements](/docs/crux/methodology/#eligibility) as the rest of CrUX, and so pages in particular may not have a complete historical record.

## Visualizing the data

> So, you may ask, why is the data shaped in this way? It wasfound that this makes it easy to plot graphs. For example, here is a graph for the p75 values for [Interaction To Next Paint (INP)](https://web.dev/inp/) for [https://web.dev](https://web.dev):

Reviewer 2: wasfound -> was found
